### PR TITLE
unstructured, community, initialize langchain-unstructured package

### PR DESCRIPTION
**Pull request by Coniferish**
#### Update (2): 
A single `UnstructuredLoader` is added to handle both local and api partitioning. This loader also handles single or multiple documents.

#### Changes in `community`:
Changes here do not affect users. In the initial process of using the SDK for the API Loaders, the Loaders in community were refactored. 
Other changes include:
The `UnstructuredBaseLoader` has a new check to see if both `mode="paged"` and `chunking_strategy="by_page"`. It also now has `Element.element_id` added to the `Document.metadata`. `UnstructuredAPIFileLoader` and `UnstructuredAPIFileIOLoader`. As such, now both directly inherit from `UnstructuredBaseLoader` and initialize their `file_path`/`file` attributes respectively and implement their own `_post_process_elements` methods.

--------
#### Update:
New SDK Loaders in a [partner package](https://python.langchain.com/v0.1/docs/contributing/integrations/#partner-package-in-langchain-repo) are introduced to prevent breaking changes for users (see discussion below).

##### TODO:
- [x] Test docstring examples
--------
- **Description:** UnstructuredAPIFileIOLoader and UnstructuredAPIFileLoader calls to the unstructured api are now made using the unstructured-client sdk.
- **New Dependencies:** unstructured-client

- [x] **Add tests and docs**: If you're adding a new integration, please include
  - [x] a test for the integration, preferably unit tests that do not rely on network access,
  - [x] update the description in `docs/docs/integrations/providers/unstructured.mdx`
- [x] **Lint and test**: Run `make format`, `make lint` and `make test` from the root of the package(s) you've modified. See contribution guidelines for more: https://python.langchain.com/docs/contributing/

Additional guidelines:
- Make sure optional dependencies are imported within a function.
- Please do not add dependencies to pyproject.toml files (even optional ones) unless they are required for unit tests.
- Most PRs should not touch more than one package.
- Changes should be backwards compatible.
- If you are adding something to community, do not re-import it in langchain.

TODO:
- [x] Update https://python.langchain.com/v0.1/docs/integrations/document_loaders/unstructured_file/#unstructured-api
  - `langchain/docs/docs/integrations/document_loaders/unstructured_file.ipynb`
  - The description here needs to indicate that users should install `unstructured-client` instead of `unstructured`. Read over closely to look for any other changes that need to be made.
- [x] Update the `lazy_load` method in `UnstructuredBaseLoader` to handle json responses from the API instead of just lists of elements.
  - This method may need to be overwritten by the API loaders instead of changing it in the `UnstructuredBaseLoader`.
- [x] Update the documentation links in the class docstrings (the Unstructured documents have moved)
- [x] Update Document.metadata to include `element_id` (see thread [here](https://unstructuredw-kbe4326.slack.com/archives/C044N0YV08G/p1718187499818419))

### CODE REVIEW

File: libs/community/tests/integration_tests/document_loaders/test_unstructured.py
Comment (Coniferish) [inline libs/community/tests/integration_tests/document_loaders/test_unstructured.py:90]: The loaders used this `file_filename` parameter I haven't seen before and don't see in the documentation ([here](https://python.langchain.com/v0.2/docs/integrations/providers/unstructured/) or [here](https://api.python.langchain.com/en/latest/document_loaders/langchain_community.document_loaders.unstructured.UnstructuredAPIFileLoader.html)). I assume I need to add logic so this param is still accepted, but please let me know if there's anything else to consider in how that should be handled. I replaced it with `metadata_filename` since that's the param we use in `unstructured`, but that's not actually a recognized [param](https://docs.unstructured.io/api-reference/api-services/api-parameters) for the sdk either.

Comment (MthwRobinson) [reply libs/community/tests/integration_tests/document_loaders/test_unstructured.py:90]: You're good to drop `file_filename`. `file_filename` was deprecated a while back in favor of `metadata_filename`.

https://github.com/Unstructured-IO/unstructured/pull/1191

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (Coniferish) [inline libs/community/langchain_community/document_loaders/unstructured.py:53]: The children either return an empty dictionary or `{"source": self.file_path}`.

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (Coniferish) [inline libs/community/langchain_community/document_loaders/unstructured.py:227]: This method was copied from the BaseLoader and updated to expect JSON/a list of dictionaries instead of `Element` objects.

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (Coniferish) [inline libs/community/langchain_community/document_loaders/unstructured.py:244]: @MthwRobinson, this `"paged"` mode seems to be equivalent to the `by_page` chunking. I thought about updating the request in the sdk if `self.mode == "paged"`, but I am wondering if there would be any issue with changing this process from post-processing like it is now. Do you think that's safe to do?

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (Coniferish) [inline libs/community/langchain_community/document_loaders/unstructured.py:240]: @MthwRobinson, can you confirm it's safe to remove this check since this implements using `unstructured-client`?

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (MthwRobinson) [reply libs/community/langchain_community/document_loaders/unstructured.py:229]: Doesn't this method get inherited from the base class? Wondering why you needed to copy it.

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (MthwRobinson) [reply libs/community/langchain_community/document_loaders/unstructured.py:242]: Yeah we can remove this check. This was for backward compatibility with a _very very old_ version of `unstructured`. Perhaps add a check on the `unstructured` version with a warning letting people know to upgrade to `unstructured>=0.14.0`.

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (MthwRobinson) [reply libs/community/langchain_community/document_loaders/unstructured.py:246]: I would leave this to avoid breaking things for people who are using it. Of course, `chunking_strategy="by_page"` is fine to pass in as well.

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (Coniferish) [inline libs/community/langchain_community/document_loaders/unstructured.py:312]: @MthwRobinson  The Loader also supports a "paged" mode. Should the docstrings be updated to mention this third mode or do we want to continue leaving it out and reference the unstructured's chunking options, which includes chunking `by_page`?

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (Coniferish) [inline libs/community/langchain_community/document_loaders/unstructured.py:113]: @MthwRobinson The Loader also supports a "paged" mode. Should the docstrings be updated to mention this third mode or do we want to continue leaving it out and reference the unstructured's chunking options, which includes chunking by_page?

Comment (Coniferish) [reply libs/community/langchain_community/document_loaders/unstructured.py:113]: I opted to leave it only mentioning the two modes and reference chunking options

File: docs/docs/integrations/providers/unstructured.mdx
Comment (MthwRobinson) [inline docs/docs/integrations/providers/unstructured.mdx:15]: Should the "install `unstructured-client`" part go below in the section you removed about `UnstructuredAPIFileIOLoader`? Don't see the whole file in the diff, but generally:
- For `UnstructuredFileLoader`, tell people to install `unstructured`
- For `UnstructuredAPIFileLoader`, tell people to install `unstructured-client`

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (MthwRobinson) [inline libs/community/langchain_community/document_loaders/unstructured.py:73]: Why this change?

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (MthwRobinson) [inline libs/community/langchain_community/document_loaders/unstructured.py:107]: Should we add "paged" into the description here?

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (MthwRobinson) [inline libs/community/langchain_community/document_loaders/unstructured.py:126]: Good catch

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (MthwRobinson) [inline libs/community/langchain_community/document_loaders/unstructured.py:154]: Why this change? Saw a similar one related to `Dict` above.

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (MthwRobinson) [inline libs/community/langchain_community/document_loaders/unstructured.py:192]: Same here, should we add "paged" to the description?

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (MthwRobinson) [inline libs/community/langchain_community/document_loaders/unstructured.py:271]: Why did you need to make this change?

**Review by MthwRobinson — CHANGES_REQUESTED** [review]
Couple of inline comments. Could you also add a blurb to the PR description with instructions on how to test locally?

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (Coniferish) [reply libs/community/langchain_community/document_loaders/unstructured.py:154]: Just following type-hinting practices I've learned from Steve.

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (Coniferish) [reply libs/community/langchain_community/document_loaders/unstructured.py:113]: @MthwRobinson, I raised the "paged" mode question here. It seemed like an somewhat irrelevant mode to support here now that we have it via chunking. It wasn't mentioned before and "paged" isn't a functionality added in this PR, so I thought it would be less confusing to tell users two ways to do the same thing. I also did not test what happens if you define both `mode="paged"` and `chunking_strategy="by_page"`

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (Coniferish) [reply libs/community/langchain_community/document_loaders/unstructured.py:271]: I had to refactor the classes a bit since the API Loaders were not inheriting directly from the BaseUnstructuredLoader and needed to implement some of their methods themselves. This was removed in that process.

File: docs/docs/integrations/providers/unstructured.mdx
Comment (Coniferish) [reply docs/docs/integrations/providers/unstructured.mdx:15]: Now that the API Loaders don't depend on `unstructured`, I wanted to add an install section for them here. The section below this wasn't completely accurate and seemed redundant once I added this part up here.
You can view the full diff/file [here](https://github.com/langchain-ai/langchain/pull/22779/commits/a172c1fe8aadbf9d1b3e5e6ca96f02dfe8841f2c)

I will update this to specifically mention the loaders, though

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (Coniferish) [reply libs/community/langchain_community/document_loaders/unstructured.py:73]: This stemmed from the this thread [here](https://unstructuredw-kbe4326.slack.com/archives/C044N0YV08G/p1718187499818419)

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (Coniferish) [reply libs/community/langchain_community/document_loaders/unstructured.py:154]: Lmk if you want me to revert these changes or if I can resolve these comments

Comment (Coniferish) [reply libs/community/langchain_community/document_loaders/unstructured.py:154]: Actually, here's the PEP for deprecating `typing.List` and other generics. 
https://peps.python.org/pep-0585/#abstract
https://docs.python.org/3/library/typing.html#deprecated-aliases

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (MthwRobinson) [reply libs/community/langchain_community/document_loaders/unstructured.py:113]: Even though paged wasn't added in this PR, the docstrings are out of date and don't include it so might as well clean that up while we're in here.

Since there are current users of the loader who use paged mode, my suggestion would be to emit a warning log letting the user know about `chunking_page="by_page"` and raise a `ValueError` is you set both `chunking_page="by_page"` and `mode="paged"`. I don't think we should remove that mode without having a deprecation warning at least for a while so users have a chance to migrate.

**Comment by Coniferish** [comment]
@ccurme, I believe this is ready for review if you or someone else with LangChain can take a look

File: docs/scripts/arxiv_references.py
Comment (MthwRobinson) [inline docs/scripts/arxiv_references.py:179]: Was this just a linting change?

Comment (Coniferish) [reply docs/scripts/arxiv_references.py:179]: Yes. Came from running `make format` or one of the other commands of theirs

**Review by efriis — COMMENTED** [review]
Howdy! Could you address merge conflicts and revert changes in unrelated docs files?

**Comment by Coniferish** [comment]
@efriis, should be good now

**Review by efriis — COMMENTED** [review]
I believe this will be a breaking change for people using the existing api connectors because they might not have `unstructured-client` installed.

Would you be interested in doing the api connectors in a partner package, so we can independently version control it to tolerate breaking changes like this?

**Comment by Coniferish** [comment]
> I believe this will be a breaking change for people using the existing api connectors because they might not have `unstructured-client` installed.
> 
> Would you be interested in doing the api connectors in a partner package, so we can independently version control it to tolerate breaking changes like this?

That's a good point. @MthwRobinson, what are your thoughts?

**Comment by MthwRobinson** [comment]
@efriis @Coniferish - Yeah I think that makes sense. That would just live under `libs/partners` and publish as an independent Python package right? If we go that direction, may be good to migrate the other Unstructured loaders there over time as well, though we can certainly start with just the new API loader that uses the client.

@Coniferish - good to go from my end if you want to start working that.

**Comment by Coniferish** [comment]
@MthwRobinson and @efriis, the code has been updated to maintain the API Loaders and add SDK Loaders as partner packages. Do we want to mark the API Loaders as deprecated in favor of the SDK Loaders? Similarly, how do we want to update the documentation?

**Comment by MthwRobinson** [comment]
@Coniferish - I'm in favor of marking the old API loaders as deprecated in favor of the new ones that use the SDK. I'd recommend logging a warning pointing people toward the new loaders.

File: libs/partners/unstructured/docs/document_loaders.ipynb
Comment (Coniferish) [inline libs/partners/unstructured/docs/document_loaders.ipynb:1]: This file was auto-generated when initializing the partner package. @efriis, would it be okay to remove this file and just depend on the existing documentation files (`docs/docs/integrations/providers/unstructured.mdx` and `docs/docs/integrations/document_loaders/unstructured_file.ipynb`)? Those have been updated to reference the SDK Loaders instead of the API Loaders.

File: libs/community/langchain_community/document_loaders/__init__.py
Comment (Coniferish) [inline libs/community/langchain_community/document_loaders/__init__.py:477]: `UnstructuredBaseLoader` was added to this and other files like it to make it accessible for import as a parent class for the SDK Loaders.

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (Coniferish) [inline libs/community/langchain_community/document_loaders/unstructured.py:213]: @MthwRobinson, is version 0.4.0 a reasonable benchmark for removing these loaders?

File: libs/partners/unstructured/langchain_unstructured/document_loaders.py
Comment (MthwRobinson) [reply libs/partners/unstructured/langchain_unstructured/document_loaders.py:90]: Yeah since we're moving to a new loader class here, I think it would make sense to make those updates.

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (MthwRobinson) [reply libs/community/langchain_community/document_loaders/unstructured.py:209]: Yeah that seems like plenty of warning!

**Comment by efriis** [comment]
Feel free to mark ready to review when it's good for another review/merge!

File: libs/community/langchain_community/document_loaders/__init__.py
Comment (Coniferish) [reply libs/community/langchain_community/document_loaders/__init__.py:477]: Good call. I'll prefix it.

File: docs/docs/integrations/providers/unstructured.mdx
Comment (Coniferish) [reply docs/docs/integrations/providers/unstructured.mdx:15]: Ah, both actually. `unstructured-client` is a dependency for `langchain-unstructured`. I'll update that and the instructions below that just reference `unstructured`

File: docs/docs/integrations/providers/unstructured.mdx
Comment (Coniferish) [reply docs/docs/integrations/providers/unstructured.mdx:26]: Good catch. I think I missed that when resolving conflicts when rebasing

File: libs/partners/unstructured/docs/document_loaders.ipynb
Comment (efriis) [reply libs/partners/unstructured/docs/document_loaders.ipynb:1]: yes! And feel free to remove the whole unstructured/docs directory. You have to move these files into the respective directories under root/docs!

**Comment by efriis** [comment]
Screenshot 2024-07-15 at 1 52 27 PM
error from docs build

**Comment by efriis** [comment]
And I believe the other CI job logs should be viewable in the "details" links above!

File: libs/community/langchain_community/document_loaders/__init__.py
Comment (Coniferish) [reply libs/community/langchain_community/document_loaders/__init__.py]: @efriis, The tests are failing because I changed this to `_UnstructuredBaseLoader` and that class is used in places I didn't anticipate--it's used in `libs/community/langchain_community/document_loaders/lakefs.py` and `libs/community/langchain_community/document_loaders/s3_file.py` for example. I'm not familiar with these loaders, but should I revert making the BaseLoader private? Happy to undo that or fix all the references so they refer to `_UnstructuredBaseLoader`. Lmk

File: docs/docs/integrations/providers/unstructured.mdx
Comment (Coniferish) [inline docs/docs/integrations/providers/unstructured.mdx:67]: I did not see an exiting example in the API documentation, so I removed this to not mislead users.

File: libs/langchain/langchain/document_loaders/unstructured.py
Comment (efriis) [inline libs/langchain/langchain/document_loaders/unstructured.py:1]: revert

File: poetry.lock
Comment (efriis) [inline poetry.lock:1]: revert

File: pyproject.toml
Comment (efriis) [inline pyproject.toml:1]: revert

File: libs/partners/unstructured/pyproject.toml
Comment (efriis) [inline libs/partners/unstructured/pyproject.toml:11]: Erick todo: add release notes link

File: libs/partners/unstructured/pyproject.toml
Comment (efriis) [inline libs/partners/unstructured/pyproject.toml:15]: make this more recent

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (efriis) [inline libs/community/langchain_community/document_loaders/unstructured.py:16]: i believe this will fail ci - you may have to type these as `Any` in community

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (efriis) [inline libs/community/langchain_community/document_loaders/unstructured.py:87]: revert - leave as `List` to avoid ci lint issues

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (efriis) [inline libs/community/langchain_community/document_loaders/unstructured.py:1]: could you summarize the changes in community? Are any of these changes breaking for existing users?

File: libs/partners/unstructured/pyproject.toml
Comment (Coniferish) [reply libs/partners/unstructured/pyproject.toml:15]: This file was generated when I initialized the partner package with `langchain-cli integration new` I believe. I don't have much experience with pyproject.toml files, but are you saying I should update this part manually?

File: libs/community/langchain_community/document_loaders/unstructured.py
Comment (Coniferish) [reply libs/community/langchain_community/document_loaders/unstructured.py:1]: 👍 Added to the PR description

**Review by efriis — APPROVED** [review]

**Pull request merged**
